### PR TITLE
Subset non truetype fonts

### PR DESCRIPTION
**Pull request by Munter**
Improve behavior when encountering fonts with non-truetype sources.

- Don't try to subset or preload fonts with no truetype sources
- Choose the a relevant truetype source if there are multiple to choose from and some of them aren't truetype

### CODE REVIEW

File: lib/transforms/subsetFonts.js [hunk 1]
```diff
@@ -218,8 +275,8 @@ async function getSubsetsForFontUsage(assetGraph, htmlAssetTextsWithProps, forma
                     subsetPromiseMap[promiseId].then(result => {
                         if (result) {
                             if (result.missingChars.length > 0) {
-                                missingCharsByFontSrc[fontUsage.props.src] =
-                                    result.missingChars.concat(missingCharsByFontSrc[fontUsage.props.src] || []);
+                                missingCharsByFontSrc[fontUsage.fontUrl] =
+                                    result.missingChars.concat(missingCharsByFontSrc[fontUsage.fontUrl] || []);
```
Comment (papandreou) [inline lib/transforms/subsetFonts.js]: Maybe it'd be more readable to use array spread:

```js
missingCharsByFontSrc[fontUsage.fontUrl] = [
    ...result.missingChars,
    ...(missingCharsByFontSrc[fontUsage.fontUrl] || [])
];
```

File: lib/transforms/subsetFonts.js [hunk 2]
```diff
@@ -249,7 +306,11 @@ async function getSubsetsForFontUsage(assetGraph, htmlAssetTextsWithProps, forma
 
         for (const item of htmlAssetTextsWithProps) {
             for (const fontUsage of item.fontUsages) {
-                if (!fontUsage.props.src.includes('//fonts.gstatic.com')) {
+                if (!fontUsage.fontUrl) {
+                    continue;
+                }
+
+                if (!fontUsage.fontUrl.includes('//fonts.gstatic.com')) {
```
Comment (papandreou) [inline lib/transforms/subsetFonts.js]: Not watertight. Maybe you could pick out `asset.hostname`?

File: lib/transforms/subsetFonts.js [hunk 3]
```diff
@@ -419,25 +480,31 @@ module.exports = ({
 
             assetGraph.eachAssetPreOrder(htmlAsset, traversalRelationQuery, asset => {
                 if (asset.type === 'Css' && asset.isLoaded) {
+                    const seenNodes = [];
```
Comment (papandreou) [inline lib/transforms/subsetFonts.js]: Use a `Set` to avoid the O(n)-ness of `Array#includes`?

File: lib/transforms/subsetFonts.js [hunk 4]
```diff
@@ -511,13 +570,15 @@ module.exports = ({
             const unsubsettedFontUsages = fontUsages
                 .filter(fontUsage => !subsetFontUsages.includes(fontUsage));
 
+            // console.log(htmlAssetTextsWithProps);
```
Comment (papandreou) [inline lib/transforms/subsetFonts.js]: Oopsie :)

File: lib/transforms/subsetFonts.js [hunk 5]
```diff
@@ -539,7 +600,7 @@ module.exports = ({
                     return htmlAsset.addRelation({
                         type: 'HtmlPreloadLink',
                         hrefType: 'rootRelative',
-                        to: assetGraph.findAssets({ url: fontUsage.props.src })[0],
+                        to: assetGraph.findAssets({ url: fontUsage.fontUrl })[0],
```
Comment (papandreou) [inline lib/transforms/subsetFonts.js]: Actually, `to: fontUsage.url` should work :)

File: test/transforms/subsetFonts.js [hunk 6]
```diff
@@ -2809,16 +2879,208 @@ describe('transforms/subsetFonts', function () {
...
+                                        to: {
+                                            isLoaded: true,
+                                            contentType: 'font/woff',
+                                            extension: '.woff'
+                                        }
+                                    }
+                                ]
+                            }
+                        },
+
```
Comment (papandreou) [inline test/transforms/subsetFonts.js]: Extra empty line

File: test/transforms/subsetFonts.js [hunk 7]
```diff
@@ -2809,16 +2879,208 @@ describe('transforms/subsetFonts', function () {
...
+    });
+
+
+    describe('with non-truetype fonts in the mix', function () {
+
+        it('should not attempt to subset non-truetype fonts', function () {
+            const warnSpy = sinon.spy().named('warn');
+
+            return new AssetGraph({root: __dirname + '/../../testdata/transforms/subsetFonts/non-truetype-font/'})
+                .on('warn', warnSpy)
```
Comment (papandreou) [inline test/transforms/subsetFonts.js]: You can safely remove the `warnSpy`, on master the AssetGraph instance will throw if a warn event is emitted and there are no listeners.

File: lib/transforms/subsetFonts.js [hunk 8]
```diff
@@ -37,25 +37,74 @@ function getGoogleFontSubsetCssUrl(fontProps, text) {
...
+
+    for (const format of formatOrder) {
+        const relation = cssFontFaceSrcRelations.find(r => r.format && r.format.toLowerCase() === format);
+
+        if (relation) {
+            return relation.to.url;
+        }
+    }
+
+    for (const extension of extensionOrder) {
```
[the remark this answers is not in the document]
Comment (Munter) [reply lib/transforms/subsetFonts.js]: This is a fallback for if someone doesn't put the `format()` part of a `src` declaration.

[the remark this answers is not in the document]
Comment (papandreou) [reply lib/transforms/subsetFonts.js]: Yeah, but shouldn't we be able to just pick the highest prioritized `relation` based on `relation.to.type`?

[the remark this answers is not in the document]
Comment (Munter) [reply lib/transforms/subsetFonts.js]: Ah. I forgot we had all the fonts as types now. Making that change

[the remark this answers is not in the document]
Comment (Munter) [reply lib/transforms/subsetFonts.js]: done

File: lib/transforms/subsetFonts.js (hunk 1 above)
Comment (Munter) [reply lib/transforms/subsetFonts.js]: Done

File: lib/transforms/subsetFonts.js (hunk 2 above)
Comment (Munter) [reply lib/transforms/subsetFonts.js]: Done

File: lib/transforms/subsetFonts.js (hunk 3 above)
Comment (Munter) [reply lib/transforms/subsetFonts.js]: Done

File: lib/transforms/subsetFonts.js (hunk 4 above)
Comment (Munter) [reply lib/transforms/subsetFonts.js]: Removed

File: lib/transforms/subsetFonts.js (hunk 5 above)
Comment (Munter) [reply lib/transforms/subsetFonts.js]: Done

File: test/transforms/subsetFonts.js (hunk 7 above)
Comment (Munter) [reply test/transforms/subsetFonts.js]: Removed

File: test/transforms/subsetFonts.js (hunk 6 above)
Comment (Munter) [reply test/transforms/subsetFonts.js]: Removed

**Pull request merged**